dask/base.py

```diff
@@ -1605,4 +1608,51 @@ def clone_key(key, seed):
     raise TypeError(f"Expected str or tuple[str, Hashable, ...]; got {key}")
 
 
+def print_on_compute(msg: str, val: T) -> T:
+    """When the value is computed, print it using distributed.print.
+
+    Note: this may affect performance, since it will change the graph being
+    computed.
+
+    Parameters
+    ----------
+    msg : str
+          Format string for printing out value. Default "%s".
+          It's also possible to use placeholders compatible with str.format(),
+          such as "{}" or "{0}", or the named placeholder "{result}"
+    val : T
+          Lazy value, including a Dask collection, Scalar, etc.
+
+    Returns
+    -------
+    result : T
+          Original value wrapped with an additional Delayed layer
+    """
+    msg = msg or "%s"
+
+    import builtins
+
+    from dask import delayed
+
+    first_key = next(flatten(val.__dask_keys__()))  # type: ignore
+    val_name = get_name_from_key(first_key)
+    dask_key_name = f"{val_name}-print"
+
+    @delayed  # type: ignore
+    def print_future(result) -> T:
+        try:
+            from distributed.worker import print
+        except ImportError:
+            print = builtins.print
+        if "{result" in msg:
+            print(msg.format(result=result))
+        elif "%" in msg:
+            print(msg % result)
+        else:
+            print(msg.format(result))
+        return result
+
+    return print_future(val, dask_key_name=dask_key_name)
```

Comment:
```suggestion
    return print_future(val)
```

Reply:
Given your code, the whole nested function seems unneeded to me. You could have obtained the same functionality with just a top-level function:

```python
@delayed
def print_on_compute(msg_or_val, val=no_default, /):
    try:
        from distributed.worker import print
    except ImportError:
        pass  # Use builtin print

    if val == no_default:
        msg, val = "%s", msg_or_val
    else:
        msg = msg_or_val

    print(msg % val)
    return val
```

However, there's the fundamental issue that this function changes the type from array/dataframe/bag/xarray to delayed. This is very bad for intermediate values.
The output Delayed object should be re-wrapped in the original collection type.

Reply:
> The output Delayed object should be re-wrapped in the original collection type.

How do you do that? you mean `from_delayed`?

Reply:
https://github.com/dask/dask/blob/1615724e22206112a6864bcd8b0d3f6afa07204e/dask/graph_manipulation.py#L396-L401

```python
rebuild, args = orig_collection.__dask_postpersist__()
orig_keys = list(flatten(orig_collection.__dask_keys__()))
if len(orig_keys) != 1:
    raise ValueError("Input collection must have exactly 1 chunk or partition")
 return rebuild( 
     output_delayed.dask, 
     *args, 
     rename={orig_keys[0]: output_delayed.key}, 
 ) 
```

Reply:
I tried that, but the print doesn't happen. What is wrong here?

```
diff --git a/dask/delayed.py b/dask/delayed.py
index 60090d627..325ec253f 100644
--- a/dask/delayed.py
+++ b/dask/delayed.py
@@ -783,13 +783,11 @@ def single_key(seq):
     return seq[0]
 
 
-@delayed
 def print_on_compute(msg_or_val, val=no_default, /):
     """When the value is computed, print it using distributed.print.
 
-    Note: this wraps the original collection into a Delayed object, thus
-    changing the original type. Additionally, multi-chunk collections
-    are collapsed into a single chunk, which may be undesirable.
+    Note: this can't be applied to multi-chunk collections, since it would
+    collapse them into a single chunk.
 
     Parameters
     ----------
@@ -801,17 +799,26 @@ def print_on_compute(msg_or_val, val=no_default, /):
     Returns
     -------
     result : T
-          Original value wrapped with a Delayed layer
+          Original value wrapped with delayed print layer
     """
-    try:
-        from distributed.worker import print
-    except ImportError:
-        pass  # Use builtin print
 
     if val is no_default:
         msg, val = "%s", msg_or_val
     else:
         msg = msg_or_val
 
-    print(msg % val)
-    return val
+    @delayed
+    def _print(value):
+        try:
+            from distributed.worker import print
+        except ImportError:
+            pass  # Use builtin print
+        print(msg % value)
+        return value
+
+    wrapped = _print(val)
+    rebuild, args = val.__dask_postpersist__()
+    orig_keys = list(flatten(val.__dask_keys__()))
+    if len(orig_keys) != 1:
+        raise ValueError("Input collection must have exactly 1 chunk or partition")
+    return rebuild(wrapped.dask, *args, rename={orig_keys[0]: wrapped.key})
```